Add support for kwargs in createChildAlgorithm

gemmaguest opened the pull request:
**Description of work.**

This PR adds support for keyword args in createChildAlgorithm to reduce the need to write boilerplate code when calling child algorithms from python workflow algorithms. This was identified as a requirement as part of the work for #32125.

This changes some core functionality so please review carefully.

Similar code in IFunctionAdapter has been consolidated to use the same supporting class.

**To test:**

Try using createChildAlgorithm from python, e.g.:
```
# Create a child algorithm with a mixture of positional and keyword args
parent_alg = AlgorithmManager.createUnmanaged('Load')
args={"XUnit": "Wavelength", "XMax": 10000.0}
child_alg = parent_alg.createChildAlgorithm('CreateSampleWorkspace', 0.0, version=1, NumBanks=4, BinWidth=100.0, **args)
child_alg.execute()
ws=child_alg.getProperty("OutputWorkspace").value
AnalysisDataService.addOrReplace("ws", ws)

# Check we get the same results running the algorithm normally
ws2=CreateSampleWorkspace(**args, NumBanks=4, BinWidth=100)
CompareWorkspaces(Workspace1='ws', Workspace2='ws2', Tolerance=1e-08, CheckSample=True, ToleranceRelErr=True, CheckAllData=True)
```

Also check the history of a workflow algorithm that uses createChildAlgorithm, e.g.
- Run `ReflectometryISISLoadAndProcess('INTER13460', 0.5)`
- Show the history of workspace `IvsQ_binned_13460`
- Tick Unroll on the top level algorithm and then click `Script to Clipboard`
- Check that the results match the following (apart from the path and the temporary workspace name):
```
LoadNexus(Filename='/path/to/INTER00013460.nxs', OutputWorkspace='TOF_INTER13460')
RenameWorkspace(InputWorkspace='TOF_INTER13460', OutputWorkspace='TOF_13460')
ReflectometryReductionOneAuto(InputWorkspace='TOF_13460', ThetaIn=0.5, MomentumTransferMin=0.0064527921283052733, MomentumTransferStep=0.047640977777016841, MomentumTransferMax=0.072643025200851155, ScaleFactor=1, OutputWorkspaceBinned='IvsQ_binned_13460', OutputWorkspace='IvsQ_13460', OutputWorkspaceWavelength='__TMP0x7fdee203f400')
GroupWorkspaces(InputWorkspaces='TOF_13460', OutputWorkspace='TOF')
```

There are no workflow algorithms that currently use the new kwargs functionality so for now the main thing is to ensure the current functionality is not broken. A subsequent PR as part of #32125 will exercise the kwargs functionality.


*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DavidFair committed:
Add overload to createChildAlgorithm with properties

gemmaguest committed:
Set properties by correct type

gemmaguest committed:
Add unit tests for kwargs of different types

gemmaguest committed:
Explicitly check python types for setting properties

Previously we were allowing the type checking to be done implicitly
which would mark int as bool etc.

gemmaguest committed:
Move python types extractor into core converters

gemmaguest committed:
Add support for property lists

Uses recursive variant so supports lists of lists etc.

gemmaguest committed:
Add unit test for list input properties

DavidFair committed:
Manually dispatch for std::vector<T> types

Boost will try to recurse down to the final level of a scalar value. In
our case we need to manually take over and dispatch for container types
using RTTI.

DavidFair committed:
Handle long/int mismatch between scalar and array props

IntArray is actually an array of longs, since we already have a long
type from Python we can preserve this until we try to set a scalar. In
the latter case we need to cast to an int for the underlying assignment
to work.

DavidFair committed:
Cleanup redundant assignments in NativeTypeExtractor

The compiler is smart enough to assign these types directly, this was a
workaround that was for the compiler not fully coercing the implicit
convertors for extractor types

DavidFair committed:
Move to generic impl to own class for dynamic dispatch

Move the dynamic dispatcher into it's own class. This makes it trivial
to use this mechanism to dynamically sort user inputs and handle them in
various different ways

DavidFair committed:
Switch IFunctionAdaptor to use common extractor

Ports IFunctionAdaptor to use the same adaptor whilst preserving the
same input limitations.

DavidFair committed:
Improve error handling and backwards compatibility

gemmaguest committed:
Avoid passing lists of mixed value types to algorithms

DavidFair commented:
I'm going to unmark this from draft as 6.3 master has been created and to get the ball rolling.

gemmaguest commented:
I need to move the release notes to 6.3 but the rest of this is still ready to review

gemmaguest committed:
Add release notes with new createChildAlgorithm functionality

StephenSmith25 commented on Framework/PythonInterface/mantid/api/src/Exports/Algorithm.cpp:
```diff
@@ -76,6 +89,87 @@ void translateCancel(const Algorithm::CancelException &exc) {
+
+  const std::array<std::string, 5> reservedNames = {"name", "startProgress", "endProgress", "enableLogging", "version"};
+
+  extractKwargs<std::string>(kwargs, reservedNames[0], name);
```
Does this let me overwrite the args with kwargs? I could then change the algorithm name to something else right, i.e.

```
dataX = [1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0]
dataY = [1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0,9.0,10.0,11.0,12.0]
args={"DataX": dataX, "DataY": dataY, "NSpec":4, 'name': "FitPeaks"}
parent_alg = AlgorithmManager.createUnmanaged('Load')
child_alg = parent_alg.createChildAlgorithm('CreateWorkspace', 0.0, version=1,**args)
child_alg.execute()
```
would begin to create the `CreateWorkspace` algorithm, but it would end up creating the fit algorithm due to the kwargs. Maybe we should only overwrite things if the optionals are empty

gemmaguest replied:
Good spot! I think we should throw an error if the user attempts to set the same property twice. This works if you try to set a kwarg twice - I get the error `got multiple values for keyword argument 'name'` - we've just not taken it into account with a mix of positional and kwargs setting the same thing.

DavidFair replied:
Hmm, I'm not sure the best way to solve this :thinking: 
Because we've made name effectively mandatory as a positional arg, by hardcoding arg 1 to a string.

Maybe we simply disable the name kwarg and emit a warning for that?

StephenSmith25 commented on Framework/PythonInterface/mantid/api/src/Exports/Algorithm.cpp:
```diff
@@ -76,6 +89,87 @@ void translateCancel(const Algorithm::CancelException &exc) {
+      : m_alg(alg), m_propName(propName) {}
+
+  void operator()(bool value) const override { setProp(value); }
+  void operator()(long value) const override { setProp(static_cast<int>(value)); }
```
What about if we explicitly pass a workspace as a kwarg, which we can do with the fit algorithm:

```
# create a workspace with a gaussian peak sitting on top of a linear (here flat) background
ws = CreateSampleWorkspace(Function="User Defined", UserDefinedFunction="name=LinearBackground, \
   A0=0.3;name=Gaussian, PeakCentre=5, Height=10, Sigma=0.7", NumBanks=1, BankPixelWidth=1, XMin=0, XMax=10, BinWidth=0.1)

# Setup the data to fit:
workspaceIndex = 0  # the spectrum with which WorkspaceIndex to fit
startX = 1      # specify fitting region
endX = 9      #

# Setup the model, here a Gaussian, to fit to data
tryCentre = '4'   # A start guess on peak centre
sigma = '1'          # A start guess on peak width
height = '8'         # A start guess on peak height
myFunc ='name=Gaussian, Height='+height+', PeakCentre='+tryCentre+', Sigma='+sigma
```

This works, 
```
# normal fit (this works)
fit_output = Fit( Function=myFunc,InputWorkspace=ws, Output='fit')
```

This doesn't work
```
args={"Function": myFunc, "InputWorkspace": ws, "Output": 'fit'}
parent_alg = AlgorithmManager.createUnmanaged('Load')
child_alg = parent_alg.createChildAlgorithm('Fit',0, 0, True, version=1, **args)
child_alg.execute()
```

gemmaguest replied:
Another good spot. @DavidFair Did we consider how to handle workspaces? I think we might just have missed this...

StephenSmith25 reviewed (commented):
The code changes look great and functions as described. I just have a few comments which I feel prevents the child algorithm call from behaving identical to the raw algorithm call.

gemmaguest commented on Framework/PythonInterface/mantid/api/src/FitFunctions/IFunctionAdapter.cpp:
```diff
@@ -26,6 +28,36 @@ using namespace boost::python;
+    // Previous existing code blindly converted any list type into a list of doubles.
+    // We now have to preserve this behaviour to maintain API compatibility as
+    // setValue only takes std::vector<double>.
+    std::vector<double> doubleVals;
```
Yes, it looks like this works, although I'm surprised it doesn't give a warning about the implicit conversion. Do you know why that is?

gemmaguest commented on scripts/Inelastic/CrystalField/fitting.py:
```diff
@@ -498,6 +498,8 @@ def FWHM(self, value):
                 if self.PhysicalProperty is not None and len(value) == len(self.Temperature) - len(self.PhysicalProperty):
                     value = value + [0] * len(self.PhysicalProperty)
+                    # Cast all types to match the first elem so we don't have mixed lists of int/doubles
+                    value = [type(value[0])(v) for v in value]
```
I'm not sure! @DavidFair do you remember if there was a reason we did it this way?

DavidFair replied:
In Crystal fields the latice configuration is passed as integers. This means we have to be careful of rounding in the crystal configuration (where 3 turns to 2.9999 in Python, then gets truncated to 2 in the int cast) and of course we can't blindly remove doubles.

This is one example of the above where we have to handle mixed list types :(

gemmaguest commented on Testing/SystemTests/tests/framework/CrystalFieldPythonInterface.py:
```diff
@@ -123,7 +123,7 @@ def runTest(self):
         # ---------------------------
 
         cf = CrystalField('Ce', 'C2v', B20=0.37737, B22=3.9770, B40=-0.031787, B42=-0.11611, B44=-0.12544,
```
The tests were failing due to lists of mixed types no longer being supported. I admit this makes me a bit uncomfortable. @DavidFair do you remember how difficult it would be to avoid this?

DavidFair replied:
In this case the breakage was intentional, as the implementation guesses if we have a float or a integer type.
The problem starts when we get a list of mixed lists, as the PropertyManager needs us to pick between floats or ints.

For this reason we now explicitly error out explaining to the user we either need a list of floats, or a list of ints. But we cannot take a mixed list due to the precision loss / chance of errors associated.

gemmaguest commented on Framework/PythonInterface/core/inc/MantidPythonInterface/core/Converters/PyNativeTypeExtractor.h:
```diff
@@ -0,0 +1,135 @@
+      applyVectorProp<std::string>(values);
+    } else {
+      // Recurse down
+      for (const auto &val : values) {
```
> Also I would (maybe there isn't a cast for it) never be able to set a `std::vector<std::vector<Type>>` property right?

Hmm I thought our intent was to support vectors of vectors but I think you're right, it will set the same property twice (or fail). @DavidFair do you agree?

DavidFair replied:
We should handle vectors of vectors, but it does the former behaviour - calls the `std::vector<T>` operator two times

If we want to detect this we should add a check to `Algorithm.cpp` as there may be cases where this is valid.
If, in the future we want two layer deep vectors we will need to extend the manual `typeID` checks to stop boost recursing down to scalar values

gemmaguest committed:
Minor tidying to avoid copy and reduce verbosity

gemmaguest committed:
Suppress compiler warning from boost

We pass the correct type of size_t to raw_function but internally it
then uses it as an int, giving a compiler warning we can do nothing
about

DavidFair committed:
Fix createChildAlg not accepting workspace types

Workspace types were not accepted, this is particularly noticable in Fit
functions. Using the already-existing Workspace registry we can
trivially add a conversion that covers all of these types

DavidFair committed:
Reject duplicated args and kwargs in CreateChildAlg

Rejects a duplicated kwargs and arg, as boost used to handle this before
we took over parsing manually using raw functions.
Now we will explicitly check for this, and throw a invalid_exception (we
couldn't find something for a TypeError though, so there is a potential
future improvement)

gemmaguest commented:
Closed by accident and now I can't reopen... I'll create a new one
